smart_open/s3.py

```diff
@@ -64,10 +68,150 @@
 END = 2
 WHENCE_CHOICES = [START, CURRENT, END]
 
+URI_EXAMPLES = (
+    's3://my_bucket/my_key',
+    's3://my_key:my_secret@my_bucket/my_key',
+    's3://my_key:my_secret@my_server:my_port@my_bucket/my_key',
+)
+
 _UPLOAD_ATTEMPTS = 6
 _SLEEP_SECONDS = 10
 
 
+def _safe_urlsplit(url):
+    """This is a hack to prevent the regular urlsplit from splitting around question marks.
+
+    A question mark (?) in a URL typically indicates the start of a
+    querystring, and the standard library's urlparse function handles the
+    querystring separately.  Unfortunately, question marks can also appear
+    _inside_ the actual URL for some schemas like S3.
+
+    Replaces question marks with newlines prior to splitting.  This is safe because:
+
+    1. The standard library's urlsplit completely ignores newlines
+    2. Raw newlines will never occur in innocuous URLs.  They are always URL-encoded.
+
+    See Also
+    --------
+    https://github.com/python/cpython/blob/3.7/Lib/urllib/parse.py
+    https://github.com/RaRe-Technologies/smart_open/issues/285
+    """
+    sr = urllib.parse.urlsplit(url.replace('?', '\n'), allow_fragments=False)
+    return urllib.parse.SplitResult(sr.scheme, sr.netloc, sr.path.replace('\n', '?'), '', '')
+
+
+def parse_uri(uri_as_string):
+    #
+    # Restrictions on bucket names and labels:
+    #
+    # - Bucket names must be at least 3 and no more than 63 characters long.
+    # - Bucket names must be a series of one or more labels.
+    # - Adjacent labels are separated by a single period (.).
+    # - Bucket names can contain lowercase letters, numbers, and hyphens.
+    # - Each label must start and end with a lowercase letter or a number.
+    #
+    # We use the above as a guide only, and do not perform any validation.  We
+    # let boto3 take care of that for us.
+    #
+    split_uri = _safe_urlsplit(uri_as_string)
+    assert split_uri.scheme in SCHEMES
+
+    port = DEFAULT_PORT
```

Comment:
what's about non-default port (for example for case when smbd use https://min.io/ with s3-like AP on non-443I)?

Reply:
https://github.com/RaRe-Technologies/smart_open/pull/408/files#diff-6d8b12a1a0fab36149bf68fc0ef770faR144

Reply:
wrong link?

Reply:
Yes, sorry. Trying again: https://github.com/RaRe-Technologies/smart_open/blob/c1be8deaf4abad08a447dd6be3436a11f4fecd08/smart_open/s3.py#L143